cli/src/update.rs

```diff
@@ -0,0 +1,209 @@
+use owo_colors::OwoColorize;
+use reqwest::Client;
+use semver::Version;
+use serde::{Deserialize, Serialize};
+
+const WUKONG_GITHUB_REPO: &str = "mindvalley/wukong-cli";
+const GITHUB_API_URL: &str = "https://api.github.com";
+
+#[derive(Debug, Serialize, Deserialize)]
+pub struct GithubLatestReleaseInfo {
+    pub tag_name: String,
+    pub url: String,
+    pub published_at: String,
+    pub html_url: String,
+}
+
+// check_for_update checks whether this wukong has had a newer release on Github
+pub async fn check_for_update() -> Result<(), WKCliError> {
+    let release_info = get_current_release_info()?;
+
+    if let Some(release_info) = release_info {
+        let last_update_checked_since =
+            compare_with_current_time(&release_info.checked_for_update_at);
+
+        if last_update_checked_since >= -(24.hours()) {
+            debug!("No need to check for update");
+            return Ok(());
+        }
+    }
+
+    debug!("Checking for update");
+
+    if let Some(latest_release_info) = get_latest_release_info(Some(GITHUB_API_URL)).await? {
+        let current_version = crate_version!().to_string();
+        let has_update = version_greater_than(&latest_release_info.version, &current_version);
+
+        if has_update {
+            debug!("New release found");
+            colored_print!(
+                "{} {} {} {}\n",
+                "A new release of wukong is available:".yellow(),
+                latest_release_info.version.cyan(),
+                "→".cyan(),
+                current_version.cyan()
+            );
+
+            colored_print!("To upgrade, run: brew upgrade wukong\n");
+            colored_print!("{}\n", latest_release_info.url.yellow());
+        } else {
+            debug!("No new release found");
+        }
+
+        update_release_info(latest_release_info)?;
+    }
+
+    Ok(())
+}
+
+fn get_current_release_info() -> Result<Option<ReleaseInfo>, WKCliError> {
+    let config = Config::load_from_default_path().map_err(|e| {
+        debug!("Error: {:?}", e);
+    });
+
+    if let Ok(config) = config {
+        return Ok(config.release_info);
+    }
+
+    Ok(None)
+}
+
+fn update_release_info(release_info: ReleaseInfo) -> Result<(), WKCliError> {
+    let config = Config::load_from_default_path();
+
+    match config {
+        Ok(mut config) => {
+            config.release_info = Some(release_info);
+
+            let _ = config.save_to_default_path().map_err(|e| {
+                debug!("Error: {:?}", e);
+            });
+        }
+        Err(e) => {
+            debug!("Error: {:?}", e);
+        }
+    };
+
+    Ok(())
+}
+
+fn version_greater_than(new_version: &str, current_version: &str) -> bool {
+    if let (Ok(new_version), Ok(current_version)) =
+        (Version::parse(new_version), Version::parse(current_version))
+    {
+        new_version > current_version
+    } else {
+        false
+    }
+}
+
+async fn get_latest_release_info(
+    github_api_url: Option<&str>,
```

Comment:
why the `github_api_url` param need to be an `Option`?

Reply:
aah, this is because i wanna make sure this function is testable. I need to pass the baseURL to this function

Reply:
but this still not explain why the param should be `Option`, it can be `&str` and still testable

Reply:
Here is the code, 
` github_api_url.unwrap_or(GITHUB_API_URL)` not sure if we can use unwrap_or if we do not use Option?

Reply:
yeah, the thing is you are not even using the `None` case, hence the question "why this is an `Option`?"